app/src/main/java/org/schabi/newpipe/local/history/HistoryRecordManager.java

```diff
@@ -101,31 +101,31 @@ public HistoryRecordManager(final Context context) {
 
         final OffsetDateTime currentTime = OffsetDateTime.now(ZoneOffset.UTC);
         return Maybe.fromCallable(() -> database.runInTransaction(() -> {
-            final long streamId = streamTable.upsert(new StreamEntity(info));
-
-            long duration = info.getDuration();
             // Duration will not exist if the item was loaded with fast mode, so fetch it if empty
-            if (duration < 0) {
-                duration = ExtractorHelper.getStreamInfo(
+            if (info.getDuration() < 0) {
+                final long duration = ExtractorHelper.getStreamInfo(
                         info.getServiceId(),
                         info.getUrl(),
                         false)
                         .subscribeOn(Schedulers.io())
                         .blockingGet()
                         .getDuration();
+                info.setDuration(duration);
             }
+            // Upsert to get a stream ID and update durations if we fetched one
+            final long streamId = streamTable.upsert(new StreamEntity(info));
```

Comment:
I'd not do it this way. Instead of `upsert`ing the previous info + the newly fetched duration, you should `upsert` the more complete info you got as a result of `ExtractorHelper.getStreamInfo()`. This would be the code:
```java
StreamInfo completeInfo = info;
// Duration will not exist if the item was loaded with fast mode, so fetch it if empty
if (completeInfo.getDuration() < 0) {
    completeInfo = ExtractorHelper.getStreamInfo(info.getServiceId(), info.getUrl(), false)
            .subscribeOn(Schedulers.io())
            .blockingGet();
}
// Upsert to get a stream ID and update durations if we fetched one
final long streamId = streamTable.upsert(new StreamEntity(completeInfo));
```
And then in the code below use `completeInfo` instead of `info`.

Reply:
I made it so that if we fetch the complete info, we save it, otherwise we save the incomplete info. I couldn't do it exactly that way, because the types don't match (`StreamInfo` versus `StreamInfoItem`), but the effect should be the same.